src/Consumer.cc

```diff
@@ -55,21 +56,34 @@ void Consumer::Init(Napi::Env env, Napi::Object exports) {
 struct MessageListenerProxyData {
   std::shared_ptr<pulsar_message_t> cMessage;
   Consumer *consumer;
+  std::function<void(void)> callback;
 
-  MessageListenerProxyData(std::shared_ptr<pulsar_message_t> cMessage, Consumer *consumer)
-      : cMessage(cMessage), consumer(consumer) {}
+  MessageListenerProxyData(std::shared_ptr<pulsar_message_t> cMessage, Consumer *consumer,
+                           std::function<void(void)> callback)
+      : cMessage(cMessage), consumer(consumer), callback(callback) {}
 };
 
 void MessageListenerProxy(Napi::Env env, Napi::Function jsCallback, MessageListenerProxyData *data) {
   Napi::Object msg = Message::NewInstance({}, data->cMessage);
   Consumer *consumer = data->consumer;
-  delete data;
```

Comment:
I think we should still delete the data. Replace this line with `std::unique_ptr<MessageListenerProxyData*> data_guard` is a common used RAII solution.

Reply:
In this PR, the memory allocation and deallocation of `MessageListenerProxyData` are all controlled by the method `MessageListener`. This make the code clearer. And same as above, the blocking call only accepts the raw pointer so we couldn't use unqie_ptr.

Reply:
We don't need to use `data_guard`, just still use the original pointer. The `unique_ptr` is used to delete the pointer once it's going out of the scope to replace deleting the pointer manually, which could be skipped by an early return.

Reply:
So from my understanding, we shouldn't `delete data` here. Right? Otherwise, it will result in double deletion.

Reply:
Yes. Your current code is good.